src/transport_performance/population/rasterpop.py

```diff
@@ -0,0 +1,707 @@
+
+        # build a colormap and add pcolormesh plot data, setting vmin and vmax
+        # to match the whole colormap range. transform to data_crs required to
+        # project onto base map.
+        plot_cmap = colormaps.get_cmap(cmap)
+        plot_data = self._xds.squeeze(axis=0).to_numpy()
+        vmin_data = np.nanmin(plot_data)
+        vmax_data = np.nanmax(plot_data)
+        ctf = ax.pcolormesh(
+            self._xds.squeeze().x.to_numpy(),
+            self._xds.squeeze().y.to_numpy(),
+            plot_data,
+            cmap=plot_cmap,
+            vmin=vmin_data,
+            vmax=vmax_data,
+            transform=data_crs,
+        )
+
+        # add a colorbar - converting format of smaller numbers to exponents
+        # modify the y label and reformat the tick axis to show min and max
+        cbar = plt.colorbar(
+            ctf,
+            ax=ax,
+            fraction=cbar_fraction,
+            pad=cbar_pad,
+            format=lambda x, _: f"{x:.0E}" if x <= 1 else f"{x:.0f}",
+        )
+        cbar.ax.set_ylabel(cbar_label, rotation=270, labelpad=20)
+        cbar.set_ticks(
+            np.concatenate(
+                [
+                    np.array([vmin_data]),
+                    cbar.get_ticks()[1:-1],
+                    np.array([vmax_data]),
+                ]
+            )
+        )
+
+        # defend against case attr_location is invalid
+        if attr_location not in ATTR_LOC.keys():
+            raise ValueError(
+                f"Unrecognised value for `attr_location` {attr_location}."
+                f"Expecting one of {list(ATTR_LOC.keys())}"
+            )
+
+        # create an attribution string and add it to the axis
+        grid_res = self._xds.rio.resolution()
+        attribution = (
+            f"Generated on: {datetime.strftime(datetime.now(), '%Y-%m-%d')}\n"
+            f"Population data: {var_attr}\n"
+            f"Grid Size: {abs(grid_res[0])}m x {abs(grid_res[1])}m\n"
+            f"Base map: {base_map_attr}"
+        )
+        ax.text(
+            ATTR_LOC[attr_location]["x"],
+            ATTR_LOC[attr_location]["y"],
+            attribution,
+            transform=ax.transAxes,
+            size=attr_font_size,
+            wrap=True,
+            fontdict={"name": "Arial", "color": "#000000"},
+            va=ATTR_LOC[attr_location]["va"],
+            ha=ATTR_LOC[attr_location]["ha"],
+        )
+
+        # use tight layout to maximise axis size of axis
+        plt.tight_layout()
+
+        # write to file if filepath is given, since there is no figure, need to
+        # get the current figure and resize it to match the axis before saving
+        if save is not None:
+            out_df = os.path.dirname(save)
+            if not os.path.exists(out_df):
+                os.mkdir(out_df)
+            fig = plt.gcf()
+            fig.set_size_inches(*figsize)
+            fig.savefig(save)
+            ax = None
```

Comment:
Applies to all `RasterPop.plot()` backends:

Feature request to implement a method for ensuring the user passes the correct file extension. In cases where the user tries to do something janky like:

`rp.plot(which=“matplotlib", save=“”somewhere/somefile.html)`